core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/rules/FieldsRule.scala

```diff
@@ -16,43 +16,46 @@
  */
 package tech.beshu.ror.accesscontrol.blocks.rules
 
-import cats.implicits._
 import cats.data.NonEmptySet
 import cats.implicits._
 import monix.eval.Task
 import tech.beshu.ror.accesscontrol.blocks.rules.FieldsRule.Settings
 import tech.beshu.ror.accesscontrol.blocks.rules.Rule.{RegularRule, RuleResult}
+import tech.beshu.ror.accesscontrol.blocks.variables.runtime.RuntimeMultiResolvableVariable
 import tech.beshu.ror.accesscontrol.blocks.{BlockContext, BlockContextUpdater}
-import tech.beshu.ror.accesscontrol.domain.DocumentField.{ADocumentField, NegatedDocumentField}
 import tech.beshu.ror.accesscontrol.domain.Header.Name
 import tech.beshu.ror.accesscontrol.domain.{DocumentField, Header}
 import tech.beshu.ror.accesscontrol.headerValues.transientFieldsToHeaderValue
 import tech.beshu.ror.accesscontrol.orders._
-import tech.beshu.ror.utils.ScalaOps._
+import tech.beshu.ror.accesscontrol.utils.RuntimeMultiResolvableVariableOps.resolveAll
+import tech.beshu.ror.utils.ScalaOps.SetOps
 
 class FieldsRule(val settings: Settings)
   extends RegularRule {
 
   override val name: Rule.Name = FieldsRule.name
 
   override def check[B <: BlockContext : BlockContextUpdater](blockContext: B): Task[RuleResult[B]] = Task {
-    if (!blockContext.requestContext.isReadOnlyRequest) RuleResult.Rejected()
-    else RuleResult.Fulfilled(blockContext.withAddedContextHeader(transientFieldsHeader))
+    if (!blockContext.requestContext.isReadOnlyRequest)
+      RuleResult.Rejected()
+    else {
+      val maybeResolvedFields = resolveAll(settings.fields.toNonEmptyList, blockContext).toSet
+      maybeResolvedFields.toNonEmptySet match {
+        case Some(resolvedFields) if !DocumentField.areDifferentAccessModesUsedSimultaneously(resolvedFields.toList) =>
```

Comment:
IMO it'd be better to recognise if two modes are mixed during parsing configuration at startup level, not at request handling level. `~@{user}` is negated field for every string resolved from user variables,  `@{user}` is regular field for every string resolved from user variables (also eg. `~john`)